File: DevExpress-Demo/DevExpress.ProductsDemo.Wpf/CS/DevExpress.ProductsDemo.Wpf/ModulesViewModels/SalesDemo/BillionStringToShortStringConverter.cs
using System;
using System.Globalization;
using System.Windows.Data;

namespace ProductsDemo {
    class BillionStringToShortStringConverter : IValueConverter {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            if(!(value is string))
                return null;
            string billionFormatString = "0,,.0M";
            string thousandsFormatString = "0,.0K";

            decimal dec;
            bool parsed = decimal.TryParse((string)value, out dec);
            if(parsed) {
                if(dec == 0)
                    return "0";
                else if(dec >= 0.1M * 1000000M)
                    return dec.ToString(billionFormatString);
                else
                    return dec.ToString(thousandsFormatString);
            } else
                return null;
        }
        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }
}

File: DevExpress-Demo/DevExpress.SalesDemo.Wpf/VB/DevExpress.SalesDemo.Wpf/Converters/BillionStringToShortStringConverter.vb
Imports System
Imports System.Globalization
Imports System.Windows.Data

Namespace DevExpress.SalesDemo.Wpf.Converters

    Friend Class BillionStringToShortStringConverter
        Implements IValueConverter

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            If Not(TypeOf value Is String) Then Return Nothing
            Dim billionFormatString As String = "0,,.0M"
            Dim thousandsFormatString As String = "0,.0K"
            Dim dec As Decimal
            Dim parsed As Boolean = Decimal.TryParse(CStr(value), dec)
            If parsed Then
                If dec = 0 Then
                    Return "0"
                ElseIf dec >= 0.1D * 1000000D Then
                    Return dec.ToString(billionFormatString)
                Else
                    Return dec.ToString(thousandsFormatString)
                End If
            Else
                Return Nothing
            End If
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class
End Namespace
